### File formats
Conversion of molecular information to computer and human interpretable information has necessitated the development of file formats for this data. The current field standard – the fastq format, traces back to a file format developed for the purpose aligning protein and nucleic acid sequence data, the [fasta format](https://www.pnas.org/doi/abs/10.1073/pnas.85.8.2444). The fasta format provided two pieces of information: sequence name, and sequence. The fastq format improves on this by providing additional information on the quality of the inferred sequence for each time it was observed ("read").  An overview and history of the file format can be found [here](https://doi.org/10.1093/nar/gkp1137).  
Because most short read data generated currently is via Illumina sequencing, most tools are primarily geared towards dealing with this format, reviewed [here](https://www.illumina.com/documents/products/technotes/technote_Q-Scores.pdf).

### Sequence quality assessment
This information allows us to do a few useful things: assess data quality, remove low quality data, and incorporate variation in data quality into models used for inferring genotypes and alleles. This rest of this note will focus on the first part.
During the generation of sequencing data, especially via any high-throughput sequencing method, some errors will exist in some reads. These errors exist at the level of individual bases, stretches of bases, and entire sequences. They can be introduced through DNA degradation, mysterious metabolites, sequencing error via machine or chemistry problems, or library preparation. We can use the tool [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) to assess and visualize these problems. It has both a GUI option, and can also be run fairly easily from the command line like this:
```
fastqc <your fastq.gz>
```
And can also be [installed using bioconda](https://anaconda.org/bioconda/fastqc)
There are many options to visualize here, but the Q score across reads, duplication level, and adapter content tend to be most important. As a general rule for each of these: You want Q above 30 as much as possible, Q above 20 is acceptable, and once you’re below Q20 your sequence is “bad” - your reads are heading into the “1 in 10 basecalls are wrong” zone, which is a lot of error. FastQC provides a helpful colour-guide here. 
![Screenshot of fastqc output.](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc.png)

For duplication level: you want duplication levels consistent with your expected coverage, roughly. If you anticipate 10X coverage and see duplication levels at that rate, great! If values are much higher than expected, it could be that there has been some technical errors. But also don’t worry too much – the genome is full of repetitive stuff, and some of it is going to be “duplicated” at really high rates. [A more complete explanation of duplication is provided here](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/8%20Duplicate%20Sequences.html)

Last – adapter content. This is where your library prepation has gone awry in some way, and has included molecular components that are necessary for the functioning of the sequencer, but that you don't actually want sequenced. There are a finite number of sequencer-specific sequences, most of which can be picked up by FastQC, and recognized and removed automatically by the trimming tool [fastp](https://onlinelibrary.wiley.com/doi/full/10.1002/imt2.107) Adapter-contamiantion will look like [this](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/small_rna_fastqc.html#M10). A bit is fine and can be removed with trimming, a lot, especially when you see it at the beginning of many of your reads, indicates your whole library preparation probably went wrong and you'll need to troubleshoot in the lab.
